subprojects/architecture-test/src/test/java/org/gradle/architecture/test/ProviderMigrationArchitectureTest.java

```diff
@@ -0,0 +1,133 @@
+                .anyMatch(method -> PropertyAccessorType.SETTER.propertyNameFor(method.getName()).equals(propertyNameFromGetter));
+        }
+    };
+
+    private static final DescribedPredicate<JavaClass> class_with_any_mutable_property = new DescribedPredicate<JavaClass>("class with any mutable property") {
+        @Override
+        public boolean apply(JavaClass input) {
+            return input.getAllMethods().stream()
+                .filter(getters::apply)
+                .anyMatch(haveSetters::apply);
+        }
+    };
+
+    private static final DescribedPredicate<JavaMethod> mutable_public_API_properties = ArchPredicates.<JavaMethod>are(public_api_methods)
+        .and(not(declaredIn(assignableTo(Task.class))))
+        .and(not(declaredIn(StartParameter.class)))
+        .and(not(declaredIn(Configuration.class)))
+        .and(not(declaredIn(FileCollection.class)))
+        .and(not(declaredIn(ConfigurableFileCollection.class)))
+        .and(are(declaredIn(class_with_any_mutable_property)))
+        .and(are(getters))
+        .and(not(annotatedWith(Inject.class)))
+        .as("mutable public API properties");
+
+    @SuppressWarnings("deprecation")
+    private static final DescribedPredicate<JavaMethod> task_properties = ArchPredicates.<JavaMethod>are(public_api_methods)
+        .and(declaredIn(assignableTo(Task.class)))
+        .and(are(getters))
+        .and(not(annotatedWith(Inject.class)))
+        .and(not(declaredIn(Task.class)))
+        .and(not(declaredIn(DefaultTask.class)))
+        .and(not(declaredIn(org.gradle.api.internal.AbstractTask.class)))
+        .as("task properties");
+
+    @ArchTest
+    public static final ArchRule mutable_public_api_properties_should_be_providers = freeze(methods()
+        .that(are(mutable_public_API_properties))
+        .and().doNotHaveRawReturnType(TextResource.class)
+        .and().doNotHaveRawReturnType(assignableTo(FileCollection.class))
+        .should().haveRawReturnType(assignableTo(Provider.class)));
+
+    @ArchTest
+    public static final ArchRule mutable_public_api_properties_should_be_file_collections = freeze(methods()
+        .that(are(mutable_public_API_properties))
+        .and().haveRawReturnType(assignableTo(FileCollection.class))
+        .should().haveRawReturnType(assignableTo(ConfigurableFileCollection.class)));
+
+    @ArchTest
+    public static final ArchRule mutable_public_api_properties_should_not_use_text_resources = freeze(methods()
+        .that(are(mutable_public_API_properties))
+        .should().notHaveRawReturnType(TextResource.class));
+
+    @ArchTest
+    public static final ArchRule public_api_task_properties_are_providers = freeze(methods()
+        .that(are(task_properties))
+        .and().doNotHaveRawReturnType(TextResource.class)
+        .and().doNotHaveRawReturnType(assignableTo(FileCollection.class))
+        .should().haveRawReturnType(assignableTo(Provider.class)));
```

Comment:
Also here:
Why did we decide on this being a `Provider` instead of a `Property`? Since they are mutable, I guess they should be `Property`.

Reply:
They are not not necessarily mutable. These are all getters in Tasks. Like the other ones are all getters in types which have any setter (not necessarily a setter for the same property). If the getter already returns a `Provider` or a subtype, I don't think we need to do anything else.

Reply:
Looking at this again, I suppose we should make the test more precise: If there is a setter for the given property, then it should return a `Property`, and if there isn't a `Provider` (or a `Property`). Same for `ConfigurableFileCollection` vs. `FileCollection`.

Reply:
I changed the assertions. If the property has a setter, then we ensure it has the mutable type (`Property`/`ConfigurableFileCollection`), if it doesn't then then immutable type is enough.